python/python/lance/fragment.py

```diff
@@ -37,53 +36,161 @@
+    @property
+    def num_deletions(self) -> int:
+        """The number of rows that have been deleted from this fragment."""
+        if self.deletion_file is None:
+            return 0
+        else:
+            return self.deletion_file.num_deleted_rows
 
-    def __eq__(self, other: object) -> bool:
-        if not isinstance(other, FragmentMetadata):
-            return False
-        return self._metadata.__eq__(other._metadata)
+    @property
+    def num_rows(self) -> int:
+        """The number of rows in this fragment after deletions."""
+        return self.physical_rows - self.num_deletions
 
-    def to_json(self) -> str:
-        """Serialize :class:`FragmentMetadata` to a JSON blob"""
-        return json.loads(self._metadata.json())
+    def data_files(self) -> List[DataFile]:
+        warnings.warn(
+            "FragmentMetadata.data_files is deprecated. Use .files instead.",
+            DeprecationWarning,
+        )
+        return self.files
+
+    def to_json(self) -> dict:
+        """Get this as a simple JSON-serializable dictionary."""
+        files = [asdict(f) for f in self.files]
+        for f in files:
+            f["path"] = f.pop("_path")
+        return dict(
+            id=self.id,
+            files=files,
+            physical_rows=self.physical_rows,
+            deletion_file=(
+                self.deletion_file.asdict() if self.deletion_file is not None else None
+            ),
+            row_id_meta=(
+                self.row_id_meta.asdict() if self.row_id_meta is not None else None
+            ),
+        )
 
     @staticmethod
     def from_json(json_data: str) -> FragmentMetadata:
-        """Reconstruct :class:`FragmentMetadata` from a JSON blob"""
-        return FragmentMetadata(json_data)
+        json_data = json.loads(json_data)
+
+        deletion_file = json_data.get("deletion_file")
+        if deletion_file is not None:
+            deletion_file = DeletionFile(**deletion_file)
+
+        row_id_meta = json_data.get("row_id_meta")
+        if row_id_meta is not None:
+            row_id_meta = RowIdMeta(**row_id_meta)
+
+        return FragmentMetadata(
+            id=json_data["id"],
+            files=[DataFile(**f) for f in json_data["files"]],
+            physical_rows=json_data["physical_rows"],
+            deletion_file=deletion_file,
+            row_id_meta=row_id_meta,
+        )
 
-    def data_files(self) -> Iterable[str]:
-        """Return the data files of the fragment"""
-        return self._metadata.data_files()
 
-    def deletion_file(self):
-        """Return the deletion file, if any"""
-        return self._metadata.deletion_file()
+@dataclass
+class DataFile:
+    """
+    A data file in a fragment.
+
+    Attributes
+    ----------
+    path : str
+        The path to the data file.
+    fields : List[int]
+        The field ids of the columns in this file.
+    column_indices : List[int]
+        The column indices in the original schema.
```

Comment:
```suggestion
    fields : List[int]
        The ids of the fields in this file.
    column_indices : List[int]
        The column indices where the fields are stored in the file.  Will  have the same length
        as `fields`.
```

Minor terminology nit (pic for pedantic reference):

![image](https://github.com/user-attachments/assets/b442c6e8-dfce-4129-9046-e6c36036e6eb)